onadata/apps/logger/data_migration/restore_backup.py

```diff
@@ -19,33 +31,75 @@ def copy_attrs(from_obj, to_obj, attrs):
 class BackupRestorer(object):
     def __init__(self, xform, version=None, restore_last=False):
         self.xform = xform
+        self.old_xml = self.xform.xml
         self.xform_backup = self._get_xform_backup(version, restore_last)
 
     def restore_xform_backup(self):
         self._restore_xform_from_backup()
+        self.migrate_surveys_to_prev_schema()
 
-        for survey in self.xform.instances.iterator():
-            try:
-                backup_survey = self.xform_backup.surveys.get(uuid=survey.uuid)
-            except BackupInstance.DoesNotExist:
-                self.migrate_survey_to_prev_schema(survey)
-            else:
-                self._restore_survey_from_backup(survey, backup_survey)
-
-    def migrate_survey_to_prev_schema(self, survey):
-        changes = self.xform_backup.changes
-        reversed_decisions = self.get_reversed_migration_decisions(changes)
+    def migrate_surveys_to_prev_schema(self):
+        changes_list = self._get_migration_changes_in_between(self.xform, self.xform_backup)
+        decisions = MigrationDecisioner.convert_changes_to_decisions(
+            self._merge_migration_changes(changes_list)
+        )
         data_migrator_factory(
-            self.xform_backup, self.xform,
-            backup_data=False, **reversed_decisions
+            self.xform_backup, self.xform, backup_data=False,
+            xml_replacement=self.old_xml, **decisions
         ).migrate()
 
+    @classmethod
+    def _merge_migration_changes(cls, changes_list):
+        empty_changes = MigrationDecisioner.construct_changes()
+        return cls._merge_migration_changes_aux(changes_list, empty_changes)
+
+    @classmethod
+    def _merge_migration_changes_aux(cls, changes_list, merged_changes):
+        if not len(changes_list):
+            return merged_changes
+        changes = changes_list.pop(0)
+        new_merged_changes = cls._merge_changes(merged_changes, changes)
+        return cls._merge_migration_changes_aux(changes_list, new_merged_changes)
```

Comment:
- issue: popping element with index 0 is linear, while popping last is O(1)
- _merge_migration_changes seems equivalent to reduce(_merge_changes, [empty_changes] + changes_list)

Reply:
Feeling ashamed that didn't spot it in a place that it is reduce. Refactored now